spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerObjectMapperImpl.java

```diff
@@ -0,0 +1,336 @@
+import com.google.cloud.spanner.ResultSet;
+import com.google.cloud.spanner.Struct;
+import com.google.cloud.spanner.Type;
+import com.google.cloud.spanner.ValueBinder;
+import com.google.common.collect.ImmutableMap;
+
+import org.springframework.data.mapping.PersistentPropertyAccessor;
+import org.springframework.data.mapping.PropertyHandler;
+import org.springframework.util.Assert;
+
+/**
+ * @author Chengyuan Zhao
+ */
+public class SpannerObjectMapperImpl implements SpannerObjectMapper {
+
+	private final SpannerMappingContext spannerMappingContext;
+
+	// used for caching the methods used for reading to entities. Does not contain the
+	// methods for
+	// properties with inner types, such as Iterable.
+	private final Map<Class, Method> propertyReadMethodMapping = new HashMap<>();
+
+	// used for caching the methods used for writing to mutations. Does not contain the
+	// methods for
+	// properties with inner types, such as Iterable.
+	private final Map<Class, Method> propertyWriteMethodMapping = new HashMap<>();
+
+	private static final Map<Class, BiConsumer<ValueBinder<WriteBuilder>, Iterable>>
+			writeBuilderIterableMapping =
+			// The casting of each biconsumer below is needed for Java 8 compilation, but not Java 9.
+			new ImmutableMap.Builder<Class, BiConsumer<ValueBinder<WriteBuilder>, Iterable>>()
+			.put(Boolean.class,
+					(BiConsumer<ValueBinder<WriteBuilder>, Iterable>) (binder,
+							value) -> binder.toBoolArray(value))
+			.put(Long.class,
+					(BiConsumer<ValueBinder<WriteBuilder>, Iterable>) (binder,
+							value) -> binder.toInt64Array(value))
+			.put(String.class,
+					(BiConsumer<ValueBinder<WriteBuilder>, Iterable>) (binder,
+							value) -> binder.toStringArray(value))
+			.put(Double.class,
+					(BiConsumer<ValueBinder<WriteBuilder>, Iterable>) (binder,
+							value) -> binder.toFloat64Array(value))
+			.put(Timestamp.class,
+					(BiConsumer<ValueBinder<WriteBuilder>, Iterable>) (binder,
+							value) -> binder.toTimestampArray(value))
+			.put(Date.class,
+					(BiConsumer<ValueBinder<WriteBuilder>, Iterable>) (binder,
+							value) -> binder.toDateArray(value))
+			.put(ByteArray.class,
+					(BiConsumer<ValueBinder<WriteBuilder>, Iterable>) (binder,
+							value) -> binder.toBytesArray(value))
+			.build();
+
+	private static final Map<Class, BiFunction<Struct, String, List>>
+			readIterableMapping =
+			new ImmutableMap.Builder<Class, BiFunction<Struct, String, List>>()
+			.put(Boolean.class, (struct, colName) -> struct.getBooleanList(colName))
+			.put(Long.class, (struct, colName) -> struct.getLongList(colName))
+			.put(String.class, (struct, colName) -> struct.getStringList(colName))
+			.put(Double.class, (struct, colName) -> struct.getDoubleList(colName))
+			.put(Timestamp.class, (struct, colName) -> struct.getTimestampList(colName))
+			.put(Date.class, (struct, colName) -> struct.getDateList(colName))
+			.put(ByteArray.class, (struct, colName) -> struct.getBytesList(colName))
```

Comment:
we can just use method reference here too

Reply:
Changes made!